# Whole Genome Long-read Sequencing Pipeline for Structural Variant Detection
### *Constructed by David Lin*
A pipeline created for long-read Oxford Nanopore Technologies (ONT) DNA whole genome sequencing data with additional comparative visualization of allele-specific telomere length (TL) estimation and structural variant (SV) calling
## Introduction

### Background

### Purpose

### Rationale

## Usage

### Installation

This pipeline requires `Conda`/`Miniconda`, `Git`, and `Nextflow`.

[`Conda`](https://docs.conda.io/projects/conda/en/stable/user-guide/install/index.html)/[`Miniconda`](https://www.anaconda.com/docs/getting-started/miniconda/install) is needed to run all the dependencies of this workflow. `Miniconda` is adequate for this pipeline. (`Conda` version 25.9.1 was used in testing, but the latest version should work.)

After you have set up `Conda` environment on your device, please install `Git` and `Nextflow` in your environment.

- [`Git`](https://git-scm.com/install) is for version control. Stay up to date with the latest release! (`Git` version 2.51.0 was used in testing, but the latest version should work.)

- [`Nextflow`](https://www.nextflow.io/docs/latest/install.html#conda) is needed to run the pipeline.  (`Nextflow` version 25.10.1 was used in testing, but the latest version should work.)

*Skip the documentations, just give me the command:*

```
conda install anaconda::git bioconda::nextflow
```

After installations, clone this repository:

```         
git clone https://github.com/davidlin5922/ontwgs_pipeline.git
```

### Dependencies

*The pipeline will automatically download and use all packages when running, so you do not need to manually download them.*

`<channel>::<package name>==<version>`
``` 
anaconda::matplotlib==3.10.6
anaconda::python=3.12
anaconda::pandas==2.3.3
bioconda::minimap2==2.30
bioconda::nanoplot==1.46.1
bioconda::samtools==1.22.1
bioconda::sniffles==2.7.1
conda-forge::r-base==4.5.2
conda-forge::r-rmarkdown==2.30
hcc::sniffles2-plot==0.2.1
```

### Inputs and Outputs

Directory Structure:

```
└── ontwgs_pipeline/
    ├── config/
    │   ├── FASTQ_paths
    │   └── sample_names
    ├── example/
    │   ├── SiHa-ONT_1000_1.fq
    │   └── SiHa-ONT_1000_2.fq
    ├── images/
    │   └── workflow.png
    ├── results/
    │   └── ...
    ├── scripts/
    │   ├── plot_bamqc.py
    │   └── report.rmd
    ├── work/
    │   └── .../
    ├── .gitignore
    ├── main.nf
    ├── nextflow.config
    ├── params.yaml
    └── README.md
```

#### Parameters

This pipeline has three mandatory parameters: `sample_names`, `FASTQ_paths`, and `reference_genome`. They are set in `params.yaml` by default:

```
sample_names: ./config/sample_names
FASTQ_paths: ./config/FASTQ_paths
reference_genome: <hg19|hg38|CHM13>
```

You need to put in unique sample names (no whitespace) in `config/sample_names` and their corresponding FASTQ filepaths in `config/FASTQ_paths`, one per line.

You must also put in the reference genome you want to use in `params.yaml`. The current available ones are `hg19`/`GRCh37`, `hg38`/`GRCh38`, `CHM13`/`T2T`/`t2t`.

#### Running the Pipeline

After you clone the repository to your device and set your parameters, you can start running the pipeline.

```         
nextflow run main.nf -params-file params.yaml
```

It will take under 10 minutes for the first run to set up all the conda environments.

#### Outputs

The output files will be generated in `results/`:

```
└── ontwgs_pipeline/
    ├── .../
    ├── results/
    │   └── bamqc/
    │   └── bams/
    │   └── qc/
    │   └── sniffles2/
    │   └── report.html
```

`bamqc`, `qc`, and `sniffles2` all contains folders named by the same_names you gave in which is where the tables and figures in `report.html` originated. `bams` contains all the aligned, sorted, and indexed BAM files for you. `sniffles2` also contains the individual and merged variant calling format (VCF) files.

### Pipeline Overview

![](images/workflow.png)

This pipeline will take in pair-end FASTQ generated by ONT-WGS and go through 4 phases:

1.  Quality Control

`nanoplot` is used to assess the quality of your raw ONT reads, including read-length and read-quality distributions, N50, and yield. Plots and a summary table are generated for each sample to provide an overview before downstream processing. The summary table of all samples are also comsolidated in the `All` tab (if more than one samples are present).

2.  Alignment

Reads are aligned to the reference genome of your choice using `minimap2` and optimized for ONT long-read data. Resulting BAM files are sorted, indexed, and passed through `samtools stats` to quantify alignment quality, mapping metrics, and read-length characteristics. A customized Python script `script/plot_bamqc.py` is used to generate the mapping quality distribution and sequencing depth computed.

3.  Structural Variant Calling

SVs are called using `Sniffles2`, which was designed for long-read data. The pipeline produces per-sample VCF files, SNF files for multi-sample merging, a combined VCF of all samples in the run, which will be used for multi-sample comparison.  Visualizations of the VCF files are built with `sniffles2-plot`.

4.  Result Visualization

All figures, tables, and summary metrics are consolidated into a single `report.html` generated via `RMarkdown`.
The report includes QC plots and metrics, alignment and mapping statistics, read-length and coverage summaries, SV distributions and comparisons across samples, and cohort-level visualizations.

### Sample Run

A toy dataset is included in `example/`. The two FASTQ files each contains 1000 reads from the SiHa cell line, isolated from fragments of a primary uterine tissue sample from a 55-year-old, female, Japanese patient with squamous cell carcinoma. The expected report (CHM13/T2T reference used) is available in `results/report.html`. You can use it to see if the pipeline runs correctly.

### Conclusion

### References
