# Whole Genome Long-read Sequencing Pipeline
### *Constructed by David Lin*
A pipeline created for long-read Oxford Nanopore Technologies (ONT) DNA whole genome sequencing data with additional comparative visualization of allele-specific telomere length (TL) estimation and structural variant (SV) calling
## Introduction

### Background

### Purpose

### Rationale

## Usage

### Installation

This pipeline requires `Conda`/`Miniconda`, `Git`, and `Nextflow`.

[`Conda`](https://docs.conda.io/projects/conda/en/stable/user-guide/install/index.html)/[`Miniconda`](https://www.anaconda.com/docs/getting-started/miniconda/install) is needed to run all the dependencies of this workflow. `Miniconda` is adequate for this pipeline. (`Conda` version 25.9.1 was used in testing, but the latest version should work.)

After you have set up `Conda` environment on your device, please install `Git` and `Nextflow` in your environment.

- [`Git`](https://git-scm.com/install) is for version control. Stay up to date with the latest release! (`Git` version 2.51.0 was used in testing, but the latest version should work.)

- [`Nextflow`](https://www.nextflow.io/docs/latest/install.html#conda) is needed to run the pipeline.  (`Nextflow` version 25.10.1 was used in testing, but the latest version should work.)

*Skip the documentations, just give me the command:*

```
conda install anaconda::git bioconda::nextflow
```

After installations, clone this repository:

```         
git clone https://github.com/davidlin5922/ontwgs_pipeline.git
```

### Dependencies

*The pipeline will automatically download and use all packages when running, so you do not need to manually download them.*

`<channel>::<package name>==<version>`
``` 
anaconda::matplotlib==3.10.6
anaconda::python=3.12
anaconda::pandas==2.3.3
bioconda::minimap2==2.30
bioconda::nanoplot==1.46.1
bioconda::samtools==1.22.1
bioconda::sniffles==2.7.1
bioconda::telogator2==2.2.2
conda-forge::r-base==4.5.2
conda-forge::r-rmarkdown==2.30
hcc::sniffles2-plot==0.2.1
```

### Inputs and Outputs

Directory Structure:

```
└── ontwgs_pipeline/
    ├── .nextflow
    ├── config/
    │   ├── fastq_paths
    │   └── sample_names
    ├── example/
    │   ├── 
    │   └── 
    ├── results/
    │   └── ...
    ├── scripts/
    │   ├── plot_bamqc.py
    │   └── report.rmd
    ├── work/
    │   └── .../
    ├── .gitignore
    ├── .nextflow.log
    ├── main.nf
    ├── nextflow.config
    ├── params.yaml
    └── README.md
```

#### Parameters

This pipeline has three mandatory parameters: `sample_names`, `fastq_paths`, and `reference_genome`. They are set in `params.yaml` by default:

```
sample_names: ./config/sample_names
fastq_paths: ./config/fastq_paths
reference_genome: <hg19|hg38|CHM13>
```

You need to put in unique sample names in `config/sample_names` and their corresponding fastq filepaths in `config/fastq_paths`, one per line.

You must also put in the reference genome you want to use in `params.yaml`. The current available ones are `hg19`/`GRCh37`, `hg38`/`GRCh38`, `CHM13`/`T2T`/`t2t`.

#### Running the Pipeline

After you clone the repository to your device and set your parameters, you can start running the pipeline.

```         
nextflow run main.nf -params-file params.yaml
```

It will take under 10 minutes for the first run to set up all the conda environments.

#### Outputs

The output files will be generated in `results/`:

```
└── ontwgs_pipeline/

    ├── results/
    │   └── 
    │   └── report.html
```

### Pipeline Overview

![](dag.png)

This pipeline will take in pair-end FastQ generated by ONT-WGS and go through 5 phases

1.  Quality Control



2.  Alignment



3.  Telomere Length Estimation



4.  Structural Variant Calling



5.  Result Visualization



### Sample Run



### Conclusion

### References
