#!/usr/bin/env rmd

---
title: "Whole Genome Long-read Sequencing Report"
output:
    html_document
params:
    sample_name: NA
    qc: NA
    bamqc: NA
    telLengths: NA
    svSummary: NA
    allSV: NA
    reference: NA
    processDir: NA
---

```{r, include=FALSE, eval=TRUE}
sv_dat = data.frame(sample_name=params$sample_name, sv_summary=paste0(params$svSummary, "/", params$sample_name))
qcs = data.frame(sample_name=params$sample_name, qc=params$qc, bamqc=params$bamqc, telLength=params$telLengths)
dat = merge(qcs, sv_dat)
```

```{r, results='asis', echo=FALSE}
## saveRDS(dat, params$processDir)
cat(paste0("**Number of sample: ", length(params$sample_name), "**  \n"))
cat(paste0("**Reference aligned: ", params$reference, "**  \n"))
# print(dat)
qc_all = matrix(nrow=23, ncol=0)
bamqc_all = matrix(nrow=39, ncol=0)
# tel_all
```

```{css, echo=FALSE}
h5 { font-size: 200%; }
img {
    display: block;
    margin-left: auto;
    margin-right: auto;
}
.main-container {
    max-width: 1800px !important;
}
```

## Select Sample and Content {.tabset}

```{r sample_tabs, results='asis', echo=FALSE}

for (i in 1:nrow(dat)) {

  dat_i <- dat[i,]

  # ----------------------------------------------------------
  # Outer tab: sample name
  # ----------------------------------------------------------
  cat(sprintf("### %s {.tabset}\n\n", dat_i$sample_name))


  # ==========================================================
  # FastQC TABSET
  # ==========================================================
  cat("#### FastQC \n\n")

  ## Basic Stats
  cat("##### Basic Statistics \n")
  nanoplot_summary = read.delim(paste0(dat_i$qc, "/NanoStats.txt"))
  colnames(nanoplot_summary) = c("Metric", "Value")
  qc_all = cbind(qc_all, nanoplot_summary[,2])
  print(knitr::kable(nanoplot_summary))

  ## QC Matrics
  # cat("##### Length vs quality\n")
  # cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$qc, "/LengthvsQualityScatterPlot_dot.png")))
  
  # cat("##### Per base sequence quality\n")
  # cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$qc, "/LengthvsQualityScatterPlot_kde.png")))
  
  cat("##### Read length distribution\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$qc, "/Non_weightedHistogramReadlength.png")))
  
  cat("##### Read length distribution (log transformed)\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$qc, "/Non_weightedLogTransformed_HistogramReadlength.png")))
  
  # cat("##### Per base sequence quality\n")
  # cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$qc, "/WeightedHistogramReadlength.png")))
  
  # cat("##### Per base sequence quality\n")
  # cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$qc, "/WeightedLogTransformed_HistogramReadlength.png")))
  
  cat("##### Yield by length\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$qc, "/Yield_By_Length.png")))
  
  # ==========================================================
  # Alignment Quality TABSET
  # ==========================================================
  cat("#### MappingQC \n\n")

  cat("##### Read length distribution\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$bamqc, "/readLength.png")))
  
  cat("##### Mapping quality distribution\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$bamqc, "/mapQuality.png")))
  
  cat("##### Sequencing depth\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$bamqc, "/depth.png")))
  
  cat("##### Alignment summary statistics\n")
  bam_summary = read.delim(paste0(dat_i$bamqc, "/summary.tsv"), header = FALSE)[,-3]
  colnames(bam_summary) = c("Metric", "Value")
  bamqc_all = cbind(bamqc_all, bam_summary[,2])
  print(knitr::kable(bam_summary))

  # ==========================================================
  # Telomere Length TABSET
  # ==========================================================
  # cat("#### Telomere Estimation \n\n")
  # cat("##### Overall read length distribution\n")
  # cat(sprintf("<img src='%s' width='100%%'>\n\n", dat_i$telLength, "/readlens.png"))

  # cat("##### Telomeric variant repeat distribution\n")
  # cat(sprintf("<img src='%s' width='100%%'>\n\n", dat_i$telLength, "/all_final_alleles.png"))

  # cat("##### Allele-specific telomere length\n")
  # cat(sprintf("<img src='%s' width='100%%'>\n\n", dat_i$telLength, "/violin_atl.png"))
  
  # ==========================================================
  # Structural Variation TABSET
  # ==========================================================
  cat("#### Structural Variation \n\n")
  
  # cat("##### Structural variant count\n")
  # cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$sv_summary, "/variant_count.jpg")))
  
  cat("##### Insertion/Deletion genotype frequncy\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$sv_summary, "/del_ins_genotype.jpg")))
  
  cat("##### Inversion/Duplication genotype frequency\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$sv_summary, "/inv_dup_genotype.jpg")))
  
  cat("##### Insertion/Deletion length distribution\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$sv_summary, "/del_ins_type_size.jpg")), "\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$sv_summary, "/length_variant.jpg")))
  
  cat("##### Inversion/Duplication length distribution\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(dat_i$sv_summary, "/dup_inv_type_size.jpg")))
  
  cat("\n---\n\n")

}

if (nrow(dat) > 1) {

  cat(sprintf("### All {.tabset}\n\n"))
  cat("#### FastQC \n\n")
  qc_all = as.data.frame(cbind(nanoplot_summary[,1], qc_all))
  colnames(qc_all) = c("Metric", dat$sample_name)
  print(knitr::kable(qc_all))
  cat("#### MappingQC \n\n")
  bamqc_all = as.data.frame(cbind(bam_summary[,1], bamqc_all))
  colnames(bamqc_all) = c("Metric", dat$sample_name)
  print(knitr::kable(bamqc_all))
  cat("#### SV Compare \n\n")
  cat("##### Heatmap\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(params$allSV, "/heatmap.jpg")))
  
  cat("##### Sample upset\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(params$allSV, "/sample_upset.jpg")))
  
  cat("##### SV site per genome\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(params$allSV, "/sv_site_per_genome.jpg")))
  
  cat("##### Variant count\n")
  cat(sprintf("<img src='%s' width='100%%'>\n\n", paste0(params$allSV, "/variant_count.jpg")), "\n")

}

```
